server/server.go

```diff
@@ -0,0 +1,325 @@
+	"github.com/v3io/frames/backends/kv"
+	"github.com/v3io/frames/v3ioutils"
+
+	"github.com/nuclio/logger"
+	"github.com/pkg/errors"
+	"github.com/valyala/fasthttp"
+)
+
+// Possible states
+const (
+	ReadyState   = "ready"
+	RunningState = "running"
+	ErrorState   = "error"
+	maxBatchSize = 10000
+)
+
+var (
+	statusPath = []byte("/_/status")
+	writePath  = []byte("/write")
+	readPath   = []byte("/read")
+)
+
+// Server is HTTP server
+type Server struct {
+	address string // listen address
+	server  *fasthttp.Server
+	state   string
+
+	//backend frames.DataBackend
+	config  *frames.V3ioConfig
+	context *frames.DataContext
+	logger  logger.Logger
+}
+
+// New creates a new server
+func New(cfg *frames.V3ioConfig, addr string, logger logger.Logger) (*Server, error) {
+	if logger == nil {
+		var err error
+		logger, err = frames.NewLogger(cfg.Verbose)
+		if err != nil {
+			return nil, errors.Wrap(err, "can't create logger")
+		}
+	}
+
+	srv := &Server{
+		address: addr,
+		state:   ReadyState,
+		config:  cfg,
+		logger:  logger,
+	}
+
+	return srv, nil
+}
+
+// State returns the server state
+func (s *Server) State() string {
+	return s.state
+}
+
+// Start starts the server
+func (s *Server) Start() error {
+	if state := s.State(); state != ReadyState {
+		s.logger.ErrorWith("start from bad state", "state", state)
+		return fmt.Errorf("bad state - %s", state)
+	}
+
+	s.server = &fasthttp.Server{
+		Handler: s.handler,
+	}
+
+	go func() {
+		err := s.server.ListenAndServe(s.address)
+		if err != nil {
+			s.logger.ErrorWith("error running HTTP server", "error", err)
+			s.state = ErrorState
+		}
+	}()
+
+	s.state = RunningState
+	s.logger.InfoWith("server started", "address", s.address)
+	return nil
+}
+
+func (s *Server) handler(ctx *fasthttp.RequestCtx) {
+	switch {
+	case bytes.Compare(ctx.Path(), statusPath) == 0:
+		fmt.Fprintf(ctx, "%s\n", s.State())
+	case bytes.Compare(ctx.Path(), writePath) == 0:
+		s.handleWrite(ctx)
+	case bytes.Compare(ctx.Path(), readPath) == 0:
+		s.handleRead(ctx)
+	default:
+		ctx.Error(fmt.Sprintf("unknown path - %q", string(ctx.Path())), http.StatusNotFound)
+	}
+}
+
+func (s *Server) handleRead(ctx *fasthttp.RequestCtx) {
+	if !ctx.IsPost() { // ctx.PostBody() blocks on GET
+		ctx.Error("unsupported method", http.StatusMethodNotAllowed)
+	}
+
+	request := &frames.ReadRequest{}
+	if err := json.Unmarshal(ctx.PostBody(), request); err != nil {
+		s.logger.ErrorWith("can't decode request", "error", err)
+		ctx.Error(fmt.Sprintf("bad request - %s", err), http.StatusBadRequest)
+		return
+	}
+
+	if request.Query != "" {
+		if err := s.populateQuery(request); err != nil {
+			s.logger.ErrorWith("can't populate query", "request", request, "error", err)
+			ctx.Error(err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	// TODO: Validate request
+
+	// TODO: We'd like to have a map of name->config of backends in configuration
+	backend, err := s.createBackend(request.Type)
+	if err != nil {
```

Comment:
@tebeka why do we create the backend every request ? we should create the backend on init

Reply:
Yeah. We talked about it and it - #7